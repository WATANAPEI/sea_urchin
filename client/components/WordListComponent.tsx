import React from "react";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import CircularProgress from "@material-ui/core/CircularProgress";
import Container from "@material-ui/core/Container";

export interface WordListComponentProps {
  wordListProps: WordListProps[];
  isLoading: boolean;
  isError: boolean;
}

export interface WordListProps {
  id: number;
  wordFront: string;
  wordBack: string;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    wordList: {
      color: "red",
      listStyle: "none"
    },
    wordListContainer: {
      display: "flex",
      flexWrap: "wrap",
      overflow: "auto",
      height: "100%"
    },
    paper: {
      padding: theme.spacing(2),
      margin: "auto"
    },
    loadingMessage: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    }
  })
);

const WordListComponent: React.FC<WordListComponentProps> = ({
  wordListProps,
  isLoading,
  isError
}: WordListComponentProps) => {
  const classes = useStyles();
  const wordLists = wordListProps.map((word: WordListProps) =>
    <Paper
      id={String(word.id)}
      key={String(word.id)}
      className={classes.paper}
    >
      <Grid container direction="row" spacing={2}>
        <Grid item>
          <h1 className="word_id">{word.id}</h1>
        </Grid>
        <Grid item xs={12} sm container>
          <Grid item xs container direction="column" spacing={2}>
            <Grid item xs>
              <h3 className="word_word">{word.wordFront}</h3>
              <h3 className="word_meaning">{word.wordBack}</h3>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </Paper>
  );
  return (
    <div className={classes.wordListContainer}>
      {isError && (
        <Container id="loadingErrorMessage" className={classes.loadingMessage}>
          Something went wrong...
        </Container>
      )}
      {!isError &&
        (isLoading ? (
          <Container id="loadingMessage" className={classes.loadingMessage}>
            <CircularProgress />
          </Container>
        ) : (
          wordLists
        ))}
    </div>
  );
};

export default WordListComponent;
