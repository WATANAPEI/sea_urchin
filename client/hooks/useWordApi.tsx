import React, { useState, useEffect } from "react";
//@ts-ignore
import { translateToWordResponse } from "./wordTranslator.ts";

export interface WordResponse {
  id: number;
  word: string;
  meaning: string;
  wordLanguageID: number;
  meaningLanguageID: number;
}

export interface ReturnData {
  word: WordResponse;
  isLoading: boolean;
  isError: boolean;
}

interface UseWordApi {
  (initialUrl: string, initialWord: WordResponse): [
    ReturnData,
    React.Dispatch<React.SetStateAction<string>>
  ];
}

const useWordApi: UseWordApi = (initialUrl, initialWord) => {
  const [word, setWord] = useState(initialWord);
  const [url, setUrl] = useState(initialUrl);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setIsError(false);
      setIsLoading(true);
      try {
        const response = await fetch(url);
        console.log(`reponse: ${JSON.stringify(response)}`);
        const word: WordResponse = await translateToWordResponse(response);
        setWord(word);
      } catch (error) {
        setIsError(true);
        console.log(`error: ${JSON.stringify(error)}`);
      }
      setIsLoading(false);
    };

    fetchData();
  }, [url]);

  return [{ word, isLoading, isError }, setUrl];
};

export default useWordApi;
