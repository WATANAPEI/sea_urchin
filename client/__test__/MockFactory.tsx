// @ts-ignore
import MockFetch from "./useWordApi.spec.tsx";
// @ts-ignore
import { WordResponse } from "../hooks/useWordApi.tsx";
// @ts-ignore
import { RawWordData, RawWordWithLangData } from "../hooks/wordTranslator.ts";
// @ts-ignore
import { WordWithLangResponse } from "../hooks/useWordsWithLangsApi.tsx";

export interface MockWordFactory {
  (condition: "success" | "failed", id: number, correctUrl: string): [
    WordResponse,
    MockFetch
  ];
}

export interface MockWordsFactory {
  (condition: "success" | "failed", id: number, correctUrl: string): [
    WordResponse[],
    MockFetch
  ];
}

export interface MockWordsWithLangsFactory {
  (condition: "success" | "failed", id: number, correctUrl: string): [
    RawWordWithLangData[],
    MockFetch
  ];
}

export interface MockFetch {
  (accessedUrl: string): Promise<Response>;
}

export const mockWordFactory: MockWordFactory = (condition, id, correctUrl) => {
  let mockWordResponse: WordResponse[];
  let mockResponse: Response;
  let mockFailedResponse: Promise<{}>;
  let mockSuccessFetch: MockFetch;
  let mockFailedFetch: MockFetch;
  const mockNotFoundResponse: WordResponse[] = [
    {
      id: -2,
      word: "Word Not Found",
      meaning: "Meaning Not Found",
      word_lang_id: -2,
      meaning_lang_id: -2
    }
  ];

  console.log(`mock ${id} is making`);
  if (condition === "success") {
    mockWordResponse = [
      {
        id: `${100 + id}`,
        word: `sucess word ${id}`,
        meaning: `success meaning ${id}`,
        word_lang_id: `${10 + id}`,
        meaning_lang_id: `${20 + id}`
      }
    ];
    console.log(`mock ${id} has mockWordResponse ${mockWordResponse[0].id}`);
    mockSuccessFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          console.log(`mockResponse succeed`);
          return await Promise.resolve({
            data: mockWordResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          console.log(`mockResponse failed`);
          return await Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      console.log(`mockSuccessFetch was called`);
      return await Promise.resolve(mockResponse);
    };
    return [mockWordResponse, mockSuccessFetch];
  } else {
    mockWordResponse = [
      {
        id: `${-100 - id}`,
        word: `failed word ${-1 * id}`,
        meaning: `failed meaning ${-1 * id}`,
        word_lang_id: `${-10 - id}`,
        meaning_lang_id: `${-20 - id}`
      }
    ];
    mockFailedFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return await Promise.resolve({
            data: mockWordResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return await Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      return await Promise.reject(mockResponse);
    };
    return [mockWordResponse, mockFailedFetch];
  }
};

export const mockWordsFactory: MockWordsFactory = (condition, id, correctUrl) => {
  let mockWordsResponse: WordResponse[] = [];
  let mockResponse: Response;
  let mockFailedResponse: Promise<{}>;
  let mockSuccessFetch: MockFetch;
  let mockFailedFetch: MockFetch;
  const mockNotFoundResponse: WordResponse[] = [
    {
      id: -2,
      word: "Word_Not_Found",
      meaning: "Meaning_Not_Found",
      word_lang_id: -2,
      meaning_lang_id: -2
    }
  ];

  //console.log(`mock ${id} is making`);
  if (condition === "success") {
    for(let i= 1; i<=10; i++){
      mockWordsResponse.push({
        id: `${100 * i + id}`,
        word: `sucess_word_${i}_${id}`,
        meaning: `success_meaning_${i}_${id}`,
        word_lang_id: `${10 * i + id}`,
        meaning_lang_id: `${20 * i + id}`
      });
    }
    //console.log(`mock ${id} has mockWordsResponse ${mockWordsResponse.id}`);
    mockSuccessFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          //console.log(`mockResponse succeed`);
          return await Promise.resolve({
            data: mockWordsResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          //console.log(`mockResponse failed`);
          return await Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      //console.log(`mockSuccessFetch was called`);
      return await Promise.resolve(mockResponse);
    };
    return [mockWordsResponse, mockSuccessFetch];
  } else {
    for( let i= 1; i <= 10; i++){
      mockWordsResponse.push({
        id: `${-100 * i - id}`,
        word: `failed_word_${i}_${id}`,
        meaning: `failed_meaning_${i}_${id}`,
        word_lang_id: `${-10 * i - id}`,
        meaning_lang_id: `${-20 * i - id}`
      });
    }
    mockFailedFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return await Promise.resolve({
            data: mockWordsResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return await Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      return await Promise.reject(mockResponse);
    };
    return [mockWordsResponse, mockFailedFetch];
  }
};

export const mockWordsWithLangsFactory: MockWordsWithLangsFactory = (condition, id, correctUrl) => {
  let mockWordsWithLangsResponse: RawWordWithLangData[] = [];
  let mockResponse: Response;
  let mockFailedResponse: Promise<{}>;
  let mockSuccessFetch: MockFetch;
  let mockFailedFetch: MockFetch;
  const mockNotFoundResponse: RawWordWithLangData[] = [
    {
      id: -2,
      word: "Word_Not_Found",
      meaning: "Meaning_Not_Found",
      word_language: "Word_Lang_Not_Fount",
      meaning_language: "Meaning_Lang_Not_Fount"
    }
  ];

  console.log(`mock ${id} is making`);
  if (condition === "success") {
    for (let i = 1; i<=10; i++) {
      mockWordsWithLangsResponse.push({
        id: `${100 * i + id}`,
        word: `sucess_word_${i}_${id}`,
        meaning: `success_meaning_${i}_${id}`,
        word_language: `sucess_word_lang_${i}_${id}`,
        meaning_language: `success_meaning_lang_${i}_${id}`
      });
    }
    mockSuccessFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          console.log(`mockResponse succeed`);
          return Promise.resolve({
            data: mockWordsWithLangsResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          console.log(`mockResponse failed`);
          return Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      console.log(`mockSuccessFetch was called`);
      return Promise.resolve(mockResponse);
    };
    return [mockWordsWithLangsResponse, mockSuccessFetch];
  } else {
    for (let i = 1; i <= 10; i++) {
      mockWordsWithLangsResponse.push({
        id: `${-100 * i - id}`,
        word: `failed_word_${i}_${id}`,
        meaning: `failed_meaning_${i}_${id}`,
        word_language: `failed_word_lang_${i}_${id}`,
        meaning_language: `failed_meaning_lang_${i}_${id}`
      });
    }
    mockFailedFetch = async (accessedUrl: string) => {
      if (accessedUrl == correctUrl) {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return Promise.resolve({
            data: mockWordsWithLangsResponse
          });
        };
      } else {
        mockResponse = new Response();
        mockResponse.json = async () => {
          return Promise.reject({
            data: mockNotFoundResponse
          });
        };
      }
      return Promise.reject(mockResponse);
    };
    return [mockWordsWithLangsResponse, mockFailedFetch];
  }
};
