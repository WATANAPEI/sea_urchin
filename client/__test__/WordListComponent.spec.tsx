import React from "react";
import { act } from "react-dom/test-utils";
// @ts-ignore
import WordListComponent, { WordListComponentProps, WordListProps } from "../components/WordListComponent.tsx";
import { createMount, createShallow } from "@material-ui/core/test-utils";
// @ts-ignore
import { WordResponse } from "../hooks/useWordApi.tsx";
// @ts-ignore
import { mockWordsFactory, MockFetch } from "./MockFactory.tsx";
import "isomorphic-fetch";
import Paper from "@material-ui/core/Paper";

describe("<WordListComponent />", () => {
  const id1 = 1;
  const stringFront1 = "this_is_front_message_1";
  const stringBack1 = "this_is_back_message_1";
  const props1 = {
    id: id1,
    wordFront: stringFront1,
    wordBack: stringBack1
  };
  const id2 = 2;
  const stringFront2 = "this_is_front_message_2";
  const stringBack2 = "this_is_back_message_2";
  const props2 = {
    id: id2,
    wordFront: stringFront2,
    wordBack: stringBack2
  };

  it("shows load message when loading", () => {
    const initialIsLoading = true;
    const initialIsError = false;
    const wordListComponentProps: WordListComponentProps = {
      wordListProps: [props1, props2],
      isLoading: initialIsLoading,
      isError: initialIsError
    };
    const wrapper = createMount()(<WordListComponent {...wordListComponentProps} />);
    expect(wrapper.find("#loadingMessage").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it("shows error message when isError is true", () => {
    const isNotLoading = false;
    const isError = true;
    const wordListComponentProps: WordListComponentProps = {
      wordListProps: [props1, props2],
      isLoading: isNotLoading,
      isError: isError
    };
    const wrapper = createMount()(<WordListComponent {...wordListComponentProps} />);
    expect(wrapper.find("#loadingErrorMessage").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it("shows WordList with 2 strings when load succeed", () => {
    const isNotLoading = false;
    const isNotError = false;
    const wordListComponentProps: WordListComponentProps = {
      wordListProps: [props1, props2],
      isLoading: isNotLoading,
      isError: isNotError
    };
    const wrapper = createShallow()(<WordListComponent {...wordListComponentProps} />);
    console.log(wrapper.debug());
    console.log(wrapper.find(".word_id"));
    expect(wrapper.find(".word_id").length).toBe(2);
    expect(wrapper.find(".word_word").length).toBe(2);
    expect(wrapper.find(".word_meaning").length).toBe(2);
  });
  it("renders properly", () => {
    const isNotLoading = false;
    const isNotError = false;
    const wordListComponentProps: WordListComponentProps = {
      wordListProps: [props1, props2],
      isLoading: isNotLoading,
      isError: isNotError
    };
    const wrapper = createShallow()(<WordListComponent {...wordListComponentProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
