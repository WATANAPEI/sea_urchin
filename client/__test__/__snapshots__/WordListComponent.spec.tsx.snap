// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WordListComponent /> renders properly 1`] = `
<div
  className="makeStyles-wordListContainer-2"
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-3"
    id="1"
    key="1"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      direction="row"
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <h1
          className="word_id"
        >
          1
        </h1>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        sm={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Grid))
          container={true}
          direction="column"
          item={true}
          spacing={2}
          xs={true}
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
            xs={true}
          >
            <h3
              className="word_word"
            >
              this_is_front_message_1
            </h3>
            <h3
              className="word_meaning"
            >
              this_is_back_message_1
            </h3>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-3"
    id="2"
    key="2"
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      direction="row"
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <h1
          className="word_id"
        >
          2
        </h1>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        sm={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Grid))
          container={true}
          direction="column"
          item={true}
          spacing={2}
          xs={true}
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
            xs={true}
          >
            <h3
              className="word_word"
            >
              this_is_front_message_2
            </h3>
            <h3
              className="word_meaning"
            >
              this_is_back_message_2
            </h3>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Paper))>
</div>
`;

exports[`<WordListComponent /> shows error message when isError is true 1`] = `
<WordListComponent
  isError={true}
  isLoading={false}
  wordListProps={
    Array [
      Object {
        "id": 1,
        "wordBack": "this_is_back_message_1",
        "wordFront": "this_is_front_message_1",
      },
      Object {
        "id": 2,
        "wordBack": "this_is_back_message_2",
        "wordFront": "this_is_front_message_2",
      },
    ]
  }
>
  <div
    className="makeStyles-wordListContainer-2"
  >
    <WithStyles(ForwardRef(Container))
      className="makeStyles-loadingMessage-4"
      id="loadingErrorMessage"
    >
      <ForwardRef(Container)
        className="makeStyles-loadingMessage-4"
        classes={
          Object {
            "fixed": "MuiContainer-fixed",
            "maxWidthLg": "MuiContainer-maxWidthLg",
            "maxWidthMd": "MuiContainer-maxWidthMd",
            "maxWidthSm": "MuiContainer-maxWidthSm",
            "maxWidthXl": "MuiContainer-maxWidthXl",
            "maxWidthXs": "MuiContainer-maxWidthXs",
            "root": "MuiContainer-root",
          }
        }
        id="loadingErrorMessage"
      >
        <div
          className="MuiContainer-root makeStyles-loadingMessage-4 MuiContainer-maxWidthLg"
          id="loadingErrorMessage"
        >
          Something went wrong...
        </div>
      </ForwardRef(Container)>
    </WithStyles(ForwardRef(Container))>
  </div>
</WordListComponent>
`;

exports[`<WordListComponent /> shows load message when loading 1`] = `
<WordListComponent
  isError={false}
  isLoading={true}
  wordListProps={
    Array [
      Object {
        "id": 1,
        "wordBack": "this_is_back_message_1",
        "wordFront": "this_is_front_message_1",
      },
      Object {
        "id": 2,
        "wordBack": "this_is_back_message_2",
        "wordFront": "this_is_front_message_2",
      },
    ]
  }
>
  <div
    className="makeStyles-wordListContainer-2"
  >
    <WithStyles(ForwardRef(Container))
      className="makeStyles-loadingMessage-4"
      id="loadingMessage"
    >
      <ForwardRef(Container)
        className="makeStyles-loadingMessage-4"
        classes={
          Object {
            "fixed": "MuiContainer-fixed",
            "maxWidthLg": "MuiContainer-maxWidthLg",
            "maxWidthMd": "MuiContainer-maxWidthMd",
            "maxWidthSm": "MuiContainer-maxWidthSm",
            "maxWidthXl": "MuiContainer-maxWidthXl",
            "maxWidthXs": "MuiContainer-maxWidthXs",
            "root": "MuiContainer-root",
          }
        }
        id="loadingMessage"
      >
        <div
          className="MuiContainer-root makeStyles-loadingMessage-4 MuiContainer-maxWidthLg"
          id="loadingMessage"
        >
          <WithStyles(ForwardRef(CircularProgress))>
            <ForwardRef(CircularProgress)
              classes={
                Object {
                  "circle": "MuiCircularProgress-circle",
                  "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                  "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                  "circleStatic": "MuiCircularProgress-circleStatic",
                  "colorPrimary": "MuiCircularProgress-colorPrimary",
                  "colorSecondary": "MuiCircularProgress-colorSecondary",
                  "indeterminate": "MuiCircularProgress-indeterminate",
                  "root": "MuiCircularProgress-root",
                  "static": "MuiCircularProgress-static",
                  "svg": "MuiCircularProgress-svg",
                }
              }
            >
              <div
                className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style={
                  Object {
                    "height": 40,
                    "width": 40,
                  }
                }
              >
                <svg
                  className="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx={44}
                    cy={44}
                    fill="none"
                    r={20.2}
                    strokeWidth={3.6}
                    style={Object {}}
                  />
                </svg>
              </div>
            </ForwardRef(CircularProgress)>
          </WithStyles(ForwardRef(CircularProgress))>
        </div>
      </ForwardRef(Container)>
    </WithStyles(ForwardRef(Container))>
  </div>
</WordListComponent>
`;
